pynest/examples/astrocytes/astrocyte_brunel_fixed_indegree.py

```diff
@@ -0,0 +1,389 @@
+       Hodgkin-Huxley like formalism. Journal of theoretical Biology, 166(4),
+       461-473. DOI: https://doi.org/10.1006/jtbi.1994.1041
+
+.. [2] De Young, G. W., & Keizer, J. (1992). A single-pool inositol
+       1,4,5-trisphosphate-receptor-based model for agonist-stimulated
+       oscillations in Ca2+ concentration. Proceedings of the National Academy
+       of Sciences, 89(20), 9895-9899. DOI:
+       https://doi.org/10.1073/pnas.89.20.9895
+
+.. [3] Nadkarni, S., & Jung, P. (2003). Spontaneous oscillations of dressed
+       neurons: a new mechanism for epilepsy?. Physical review letters, 91(26),
+       268101. DOI: https://doi.org/10.1103/PhysRevLett.91.268101
+
+See Also
+~~~~~~~~
+
+:doc:`astrocyte_small_network`
+
+"""
+
+###############################################################################
+# Import all necessary modules for simulation and plotting.
+
+import random
+
+import matplotlib.pyplot as plt
+import nest
+import numpy as np
+
+###############################################################################
+# Set simulation parameters.
+
+sim_params = {
+    "dt": 0.1,  # simulation resolution in ms
+    "pre_sim_time": 100.0,  # pre-simulation time in ms (data not recorded)
+    "sim_time": 1000.0,  # simulation time in ms
+    "N_rec_spk": 100,  # number of neurons to record from with spike recorder
+    "N_rec_mm": 50,  # number of nodes (neurons, astrocytes) to record from with multimeter
+    "n_threads": 4,  # number of threads for NEST
+    "seed": 100,  # seed for the random module
+}
+
+###############################################################################
+# Set network parameters.
+
+network_params = {
+    "N_ex": 8000,  # number of excitatory neurons
+    "N_in": 2000,  # number of inhibitory neurons
+    "N_astro": 10000,  # number of astrocytes
+    "CE": 800,
+    "CI": 200,
+    "p_third_if_primary": 0.5,  # probability of each created neuron-neuron connection to be paired with one astrocyte
+    "pool_size": 10,  # astrocyte pool size for each target neuron
+    "pool_type": "random",  # astrocyte pool will be chosen randomly for each target neuron
+    "poisson_rate": 2000,  # Poisson input rate for neurons
+}
+
+syn_params = {
+    "w_a2n": 0.01,  # weight of astrocyte-to-neuron connection
+    "w_e": 1.0,  # weight of excitatory connection in nS
+    "w_i": -4.0,  # weight of inhibitory connection in nS
+    "d_e": 2.0,  # delay of excitatory connection in ms
+    "d_i": 1.0,  # delay of inhibitory connection in ms
+}
+
+###############################################################################
+# Set astrocyte parameters.
+
+astrocyte_model = "astrocyte_lr_1994"
+astrocyte_params = {
+    "IP3": 0.4,  # IP3 initial value in µM
+    "delta_IP3": 0.5,  # Parameter determining the increase in astrocytic IP3 concentration induced by synaptic input
+    "tau_IP3": 2.0,  # Time constant of the exponential decay of astrocytic IP3
```

Comment:
@IiroAhokainen I know too little about astrocyte dynamics to be the judge on this. Maybe @HanjiaJiang could comment or colleagues in your group?

Reply:
I think these can be reconsidered indeed. The present values are just those used for our benchmarks. But at the moment I also cannot suggest what values are better (e.g., both biologically and for the purpose of demonstration). Following Iiro's comment, maybe using something similar to the default values could still work. Would it be possible for Iiro to try and suggest some values?

Reply:
Okay, I got some ideas. I will be touch by email

Reply:
After going through some parameter testing, I think the default delta_IP3=5.0 µM value in astrocyte_lr_1994.cpp is not reasonable. I think it is a forgotten artifact from the performance tests.

From Nadkarni&Jung paper: "The parameter rIP3 determines the production of IP3 in response to a neuronal action potential. The production term is activated when the membrane potential of the neuron is larger than +50 mV via the step function."


In practice, the width of a spike is around 1-2 ms. For the rate parameter, they used value 0.2 µM/s and 0.8 µM/s. Thus, in terms of quantal release, this would be 0.0002 µM or 0.0008 µM per spike (assuming the width 1 ms). 

 Therefore, I'd suggest to change the default value from 5.0 µM to 0.0002 µM following Nadkarni&Jung. 

For the examples astrocyte_brunel_fixed_indegree.py and astrocyte_brunel_bernoulli.py this would mean that we could just use the default values. 

For the examples astrocyte_single.py, astrocyte_interaction.py and astrocyte_small_network.py we could then use values e.g.
astrocyte_interaction.py: delta_IP3=2.0 µM, tau_IP3=default
astrocyte_small_network.py: delta_IP3=2.0 µM, tau_IP3=10.0 ms
astrocyte_single.py: delta_IP3=default, tau_IP3=default

Here I assumed that the w_pre2astro is set to 1.0. Of course, changing w_pre2astro would also affect the total IP3  release per spike.

Does @jugoslavaacimovic have any comments on this topic? Do you think my reasoning is reasonable?

Reply:
I will summarize our earlier discussions on this topic here. Indeed, in benchmarking tests we used larger values for delta_IP3 and smaller values for tau_IP3 than necessary. In those tests, it was not necessary to have physiologically realistic values but it was necessary to produce large calcium response in astrocytes, in order to be able to test the new code and simulator performance.
It makes sense to set default values as close as possible Nadkarni2003 paper, these values have been used a lot in computational papers and modelers know and recognize them. However, as discussed earlier, we do not use the same model as Nadkarni since we simulate dynamics of spiking network while Nadkarni had a single HH-type neuron, which might affect the choice of model parameters.

My preference would be the following: set NEST default values as close as possible to Nadkarni2003, but make sure that those default values, when placed into astrocyte_brunel model (that is part of documentation), induce the right model dynamics. The model in documentation should have a) spiking neurons, b) elevated calcium in astrocytes, c) SIC inputs to  postsynaptic neurons. The goal of the model is to illustrate neuronal and astrocyte activity and the implemented neuron-astrocyte interaction, so these mechanisms should be visible in the simulation results.

Reply:
You could set tau_IP3 to Nadkarni's value (about 7000ms) and test if the proposed value for delta_IP3 works. If yes, use it. If no, you might need to increase it a bit until the desired model activity emerges.

Reply:
I tested the Bernoulli version of model with default tau_IP3 (7142 ms) and delta_IP3=0.0002 µM. The calcium crosses the threshold and then produces a similar SIC (>20 pA) in less than 1 s already. The result is also similar, if we further leave the initial IP3 as the default (0.16 µM); just that the calcium and SIC will rise slower (at around t=2 s). Perhaps Iiro can try to reproduce this.

Reply:
As @HanjiaJiang, I tested the Bernoulli version of model with default tau_IP3 (7142 ms) and delta_IP3=0.0002 µM and changed w_a2n=0.05. These parameter values show great dynamics. As discussed, I think delta_IP3=0.0002 is as close as possible to the original value used in Nadkarni2003 paper. I also revisited the other examples.

As a conclusion, I would do the following changes:

1. Changing the default value of delta_IP3 to 0.0002.
2. For the brunel network examples, use the default astrocyte parameters and set w_a2n=0.05. 
3.  For the astrocyte_single example, use poisson_weight=1.0, delta_IP3=0.2 and tau_IP3=default.
4. For the astrocyte_interaction example, use delta_IP3=0.2 and tau_IP3=default.
5. For the astrocyte_small_network example, use delta_IP3=0.2 and tau_IP3=default.

Reply:
@jugoslavaacimovic and @HanjiaJiang Would you agree with @IiroAhokainen's suggestions? Then I would implement them.

Reply:
Hi, I noticed that, in the Brunel network, if we use default for all parameters (including delta_IP3=0.0002), the SIC is not generated within one second. But if we specify the initial IP3 value as in the previous version ({"IP3": 0.4}), then the SIC is generated at around t=500 ms already. Otherwise, we will have to extend the model time from the current 1.1 s to >2 s, in order to show a SIC. I think the former (changing IP3) is better, because it saves some simulation time for the user, and avoids showing an empty SIC for a significant amount of time. @IiroAhokainen could you try it and let me know if you agree with this?

In addition, I saw that you raised the w_a2n to 0.05. Is it simply because the SIC amplitude is more realistic this way?

Otherwise, the suggestions look good to me. Thanks a lot.

Reply:
Hi all, I would be happy with most parameter choices that produce good illustration of model dynamics. However, it is difficult to judge without seeing the simulation results. @IiroAhokainen Any chance to share your results via email? No fancy figures that require much work, just basic raster plots and calcium traces. Thanks!

Reply:
Yes, I think keeping the initial value IP3=0.4 makes sense as Han-Jia mentioned. And yes, I think the amplitude of SIC is more realistic with w_a2n=0.05 given the current parameters. 

I will send some figures via email.

Also, in the brunel examples, the pre-simulation is a bit confusing. I think the data is still recorded from the initial 100 ms. I think the recorders should be cleaned after the first simulation call if we do not want to plot the data from the first 100 ms. Am I right, or am I missing something?

Reply:
That is actually a good point, thanks. In the beginning I added the pre-simulation simply because it was a way to avoid initial transients. However, I did not notice the fact that we do not have any further analysis like synchrony and hence do not really need this. Here we only visualize the dynamics to the users, and there is probably no point to cut out data. So, I would say the pre-simulation can be discarded to avoid confusion.

Reply:
@IiroAhokainen's figures (sent via email) look very good to me. In his simulations, SIC emerges at about 500 ms simulation time although delta_IP3=0.0002. Could @HanjiaJiang comment the figures in the mail. If these are fine for you, I think we can use the parameters proposed by Iiro.

@IiroAhokainen I agree that the pre-simulation is not needed in these examples. As @HanjiaJiang explained, we need it  elsewhere, to be able to remove transition period and to consider only the steady state of the model. @HanjiaJiang's example code should remove the pre-simulation segment correctly, without any additional cleaning of recorders. At least in the version of the code that I have everything works fine. 

In these examples, the simulation time should be kept short, t_sim = 1000 ms is good, please do not extend it. The code should execute fast, otherwise it might look broken or stuck.

Reply:
> As @HanjiaJiang, I tested the Bernoulli version of model with default tau_IP3 (7142 ms) and delta_IP3=0.0002 µM and changed w_a2n=0.05. These parameter values show great dynamics. As discussed, I think delta_IP3=0.0002 is as close as possible to the original value used in Nadkarni2003 paper. I also revisited the other examples.
> 
> As a conclusion, I would do the following changes:
> 
> 1. Changing the default value of delta_IP3 to 0.0002.
> 2. For the brunel network examples, use the default astrocyte parameters and set w_a2n=0.05.
> 3. For the astrocyte_single example, use poisson_weight=1.0, delta_IP3=0.2 and tau_IP3=default.
> 4. For the astrocyte_interaction example, use delta_IP3=0.2 and tau_IP3=default.
> 5. For the astrocyte_small_network example, use delta_IP3=0.2 and tau_IP3=default.

Hi, the data all look good to me, so I will only wrap it up here. @heplesser we have agreed on Iiro's suggestions (quoted above). Plus, we have agreed on two additional points:

1. Keep using {"IP3": 0.4} for astrocytes in astrocyte_brunel_bernoulli.py and astrocyte_brunel_fixed_indegree.py.

2. Discard the code for pre-simulation in astrocyte_brunel_bernoulli.py and astrocyte_brunel_fixed_indegree.py.

@jugoslavaacimovic and @IiroAhokainen please give a thumbs-up if this looks good, or add anything you think is missing. @heplesser please make changes accordingly and let me know if there is anything I can help. Thanks!

Reply:
Looks good. Thanks everyone.

Reply:
@heplesser I think there is one additional little thing: the output figure of astrocyte_interaction.py is used for NEST documentation: doc/htmldoc/static/img/astrocyte_interaction.png. Here I attach the new version of this figure with Iiro's parameters. Could you perhaps use this to replace the old version? Thanks.
![astrocyte_interaction](https://github.com/user-attachments/assets/0311c0d9-c9f0-4d55-ad13-9b48d55181a4)

Reply:
Is in place, see 2e65b02d23e53f08bee39598517b6de0f26213ef